### ISSUE TITLE
Add a way to ask for a storage proof of just the hash of a value

### ISSUE THREAD

**Issue by tomaka** [paritytech/substrate#10623]
In the context of smoldot, I would like to store the latest runtime code in a cache on disk, and, when smoldot initializes, compare the runtime code in cache with the runtime code on chain.
The objective is to save bandwidth, as smoldot wouldn't have to download the full runtime code again if it already has it in cache.
cc https://github.com/paritytech/smoldot/issues/1769

In order for this to be useful, the light client needs to be able to ask a full node for just the hash of `:code`.
As far as I know, this isn't possible with Substrate right now, and it would be nice to add this possibility.

Unfortunately I'm not really up to date with the various trie and trie proof formats so I don't really know what to suggest.

cc @cheme

**Comment by bkchr** [comment]
That should be doable with https://github.com/paritytech/substrate/pull/9732. 

At least from the proof side, but this will still require some code path to only include the hash and not the data.

@cheme can correct me if I'm wrong. I think it should work and it should also work the way you want it for `:code`, but only "by accident". The point is that some data could be stored inside a node (because it isn't that big), but then we don't have the hash of the data in the trie. But since #9732 we are putting big data (bigger than the size of a hash) into its own "bucket" under the hash of the data. As the data in our case `:code` is probably always bigger than the size of a hash, it should work the way you want it to have.

**Comment by cheme** [comment]
:+1: yes this is only doable if working on a migrated #9732 state (not on old/current state).

There is no specific api to query the hash of the value directly, but just producing the proof is the same as running one iteration of the key iterator for prefix ':code'.

If data is less than 33 byte, we would still have a proof of its hash (by hashing the value), but to read the proof a new `get_value_hash` function on the trie would be cleaner (then the function could also be use to create the proof).

**Comment by bkchr** [comment]
> If data is less than 33 byte, we would still have a proof of its hash (by hashing the value)

Ahh you mean we would have a proof with the data in it, but we can then just request the hash of this that is being found in the proof?

**Comment by cheme** [comment]
> > If data is less than 33 byte, we would still have a proof of its hash (by hashing the value)
> 
> Ahh you mean we would have a proof with the data in it, but we can then just request the hash of this that is being found in the proof?

yes.

Also I remember that without #9732, we can also use a trie proof as implemented in https://github.com/paritytech/trie/tree/master/trie-db/src/proof : basically we remove the value from the proof and pass it as a parameter during verification (so it is more costly to check (need to hash the wasm big trie node)).
If verification succeeds we know value did not change, if it fails we know it did probably change (could also be change of node structure).

**Comment by tomaka** [comment]
https://github.com/paritytech/substrate/pull/9732 has been merged, but none of the main relay chains has been updated to `state_version = 2` yet, is that correct?

Assuming that we have a migrated state, we just need to add a new boolean flag to the request in the network protocol, so that the proof in the response doesn't include the unhashed values.

**Comment by cheme** [comment]
> https://github.com/paritytech/substrate/pull/9732 has been merged, but none of the main relay chains has been updated to state_version = 2 yet, is that correct?

yes (update is 0 to 1), you can follow https://github.com/paritytech/devops/issues/1508

I did not think about mixing hash only and value request, does not sounds like a bad idea. Maybe I will still prefer have global setting for compactness of query, but it is only because I don't expect user to mix both query in practice. But can be done this way.

**Comment by tomaka** [comment]
I don't actually understand why https://github.com/paritytech/substrate/pull/9732 is needed.
Can we not generate and send back on the network a Merkle proof that contains hashes, even when the storage itself doesn't contain these hashes?

**Comment by bkchr** [comment]
The problem for now is that if you want to proof that the `hash` of some data is part of the trie isn't that easy. We don't store the hash of the data in the trie directly. We could return a storage proof that a certain leaf is part of the trie. So, you would need to build the leaf, hash the leaf and then you could check that the leaf is part of the trie. However, when the data is for example inside a branch node, this gets even more complicated, because you would need to reconstruct the branch node, hash it etc.

**Comment by tomaka** [comment]
I don't understand your comment.

What I want is ask a full node: what is the hash of `:code`?

The full node would then return a storage proof.
A storage proof normally contains trie nodes, and in the case of leaf nodes they contain the storage item of the node, either hashed or unhashed.

I suppose that right now the implementation just copy-pastes the content of the trie into the storage proof.
However, I don't see why we can't then modify the storage proof to always contain the hashed value (if the request only desires the hashed value), no matter what the actual trie says.

**Comment by tomaka** [comment]
Ok I understood now.
If you modify the leaf nodes to contain the hashed value, you modify the hash of the leaf node. There's no way to prove that the proof is correct without including the unhashed value in it.

**Comment by bkchr** [comment]
Exactly :+1: In the future the leaf/branch nodes will directly contain the hash and we can provide this proof easily.

**Comment by tomaka** [comment]
Another use case for having only hashes in the storage proofs is when you want to implement `state_getKeysPaged`.
This JSON-RPC function returns the list of all keys starting with a certain prefix.
This is used for example to get the list of validators, I believe.

Having the values in the proof adds a lot of unnecessary bloat.

As a side note, it would also be great to add a networking protocol that asks for a proof containing all keys starting with a certain prefix, but I don't think it's worth a separate issue.

**Comment by tomaka** [comment]
Coming back. I still don't think that this is blocked on the deployment on https://github.com/paritytech/substrate/pull/9732.

What we can do when receiving a request for a "proof without values" is:

- If the chain uses `state_version = 0`, then send the values anyway.
- If the chain uses `state_version = 1`, then don't send the values.

The format of the nodes in the proof lets us know the difference between the two situations so there's no ambiguity.

I think we should implement it this way because we should preferably support this protocol even if `state_version` is still 0 or if the chain is in a hybrid state.

**Comment by tomaka** [comment]
Would it be possible to start implementing this, at least the necessary trie/client changes?

**Comment by cheme** [comment]
I did implement it in the past, but my branches are not really up to date.
I can plan to do it start of next week, I mean if nobody else plan to do it, I will do it start of next week :)

**Comment by tomaka** [comment]
Since I don't have access to https://github.com/paritytech/devops/issues/1508, can I ask what is the current progress of this update, now that 3 months have passed?

**Comment by cheme** [comment]
We did migrate westend (you did see it in the recent westend issue), I will prepare kusama and polkadot next (probably will wait to check that there is no other issue in westend and yours is fixed).
